### PR TITLE
HIVE-24723: Use ExecutorService in TezSessionPool

### PR DESCRIPTION
**Pull request by belugabehr**

### CODE REVIEW

**Comment by belugabehr** [comment]
@pgaref FYI

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java [hunk 1]
```diff
@@ -395,20 +379,38 @@ int getInitialSize() {
...
-    // This is an async method, so always launch threads, even for a single task.
-    for (int i = 0; i < threadCount; ++i) {
-      ListenableFutureTask<Boolean> task = ListenableFutureTask.create(
-          new CreateSessionsRunnable(deltaRemaining));
-      new Thread(task, "Tez pool resize " + i).start();
-      threadTasks.add(task);
+    LOG.info("Resizing the pool; adding {} sessions", toStart);
+
+    final int threadCount =
+        Math.min(toStart, HiveConf.getIntVar(initConf, ConfVars.HIVE_SERVER2_TEZ_SESSION_MAX_INIT_THREADS));
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java:385]: Shall we rename toStart -> initialSize for consistency?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java [hunk 2]
```diff
@@ -395,20 +379,38 @@ int getInitialSize() {
...
+
+    return createSessions(toStart, threadCount);
+  }
+
+  private ListenableFuture<List<Boolean>> createSessions(int sessionCount, int maxParallel) {
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java:390]: Shall we use same vars here? Add method doc?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java [hunk 3]
```diff
@@ -395,20 +379,38 @@ int getInitialSize() {
...
+    Preconditions.checkArgument(sessionCount > 0);
+    Preconditions.checkArgument(maxParallel > 0);
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java:392]: Second Precondition check is redundant

**Comment by pgaref** [comment]
There are also some test timeouts -- are they related to this TezSessionPool change? cc @abstractdog

**Comment by belugabehr** [comment]
@pgaref Ya, tests failing were from this change.  I made a typo.  Will address your comments.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java (hunk 1 above)
Comment (belugabehr) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java:385]: I would advise against that.  The first one initialized the pool, the second one is to increase the existing pool.  I think the naming is OK.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java (hunk 2 above)
Comment (belugabehr) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java:390]: Sure

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java (hunk 3 above)
Comment (belugabehr) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPool.java:392]: This was previous checked in the calling methods.  I have moved it here to remove redundancy.  Not an expensive operation and it's not called all that often, so no worries even if it is doubled somewhere.  Better safe than sorry as the code changes.

**Comment by belugabehr** [comment]
@pgaref  OK.  I got the current implementation paired down quite a bit.  Please review.

**Comment by pgaref** [comment]
> @pgaref OK. I got the current implementation paired down quite a bit. Please review.

Thanks @belugabehr changes LGTM.
@abstractdog  can you also take a look please?

**Comment by belugabehr** [comment]
@abstractdog If you have any cycles, can you please review this PR please? Thanks!

**Comment by belugabehr** [comment]
@pgaref Are you OK with your 'LGTM' or would you like @abstractdog to weigh in?

**Comment by pgaref** [comment]
> @pgaref Are you OK with your 'LGTM' or would you like @abstractdog to weigh in?

Hey @belugabehr -- I would like @abstractdog to check this out so lets give it a couple of days more. I will take another look if he is out of spare cycles :)

**Comment by abstractdog** [comment]
> > @pgaref Are you OK with your 'LGTM' or would you like @abstractdog to weigh in?
> 
> Hey @belugabehr -- I would like @abstractdog to check this out so lets give it a couple of days more. I will take another look if he is out of spare cycles :)

basically, it looks good to me, how have you tested this?
btw, I'm running some basic tests on cluster to ensure correct behavior

**Comment by belugabehr** [comment]
@abstractdog Just the existing unit tests.  I have some confidence in the unit testing because they did indeed fail on a couple of mistakes I made along the way.

**Comment by abstractdog** [comment]
> @abstractdog Just the existing unit tests. I have some confidence in the unit testing because they did indeed fail on a couple of mistakes I made along the way.

yeah, I haven't found issues, LGTM, I think it's safe to commit

**Review by abstractdog — APPROVED** [review]

**Comment by belugabehr** [comment]
@abstractdog @pgaref Hey, thanks so much for your time.  I know it's not the most exciting work ever to review, but I hope that this will slowly improve the project and provide good examples for others to follow.

**Comment by pgaref** [comment]
Thanks for the PR @belugabehr and the time to polish this! :)

**Pull request merged**